climakitae/util/utils.py

```diff
@@ -1183,3 +1185,82 @@ def stack_sims_across_locs(ds):
         for sim_name in ds["simulation"]
     ]
     return ds
+
+
+def clip_to_shapefile(
+    data: xr.Dataset | xr.DataArray,
+    shapefile_path: str,
+    feature: tuple[str, str | int | float | bool | list[str | int | float | bool]] = (),
+    name: str = "user-defined",
+    **kwargs,
+) -> xr.Dataset | xr.DataArray:
+    """Use a shapefile to select an area subset of AE data.
+
+    By default, this function will clip the data to the area covered by all features in
+    the shapefile. To clip to specific features, use the `feature` keyword.
+
+    Parameters
+    ----------
+    data : xr.Dataset | xr.DataArray
+        Data to be clipped.
+    shapefile_path: str
+        Filepath to shapefile. Shapefile must include valid CRS.
+    feature: tuple(str, str | int | float | list)
+        Tuple containing attribute name and value(s) for target feature(s) (optional).
+    name: str
+        Location name to record in data attributes if 'feature' parameter is not set (optional).
+    **kwargs
+        Additional arguments to pass to the rioxarray clip function
+
+    Returns
+    -------
+    clipped : xr.Dataset | xr.DataArray
+        Returns same type as 'data', but grid is clipped to shapefile feature(s).
+    """
+    if data.rio.crs is None:
+        raise RuntimeError(
+            "No CRS found on input parameter 'data'. Use rioxarray write_crs() method to set CRS."
+        )
+
+    region = gpd.read_file(shapefile_path)
+
+    if region.crs is None:
+        raise RuntimeError(
+            "No CRS found on data read from shapefile. Verify that shapefile contains valid CRS information."
+        )
+
+    # Select only user provided feature
+    if feature:
+        try:
+            print("Selecting feature", feature)
+            if isinstance(feature[1], list):
+                region = region[region[feature[0]].isin(feature[1])]
+            else:
+                region = region[region[feature[0]] == feature[1]]
+            if len(region) == 0:  # No features found
+                raise Exception
+        except Exception as err:
```

Comment:
I would raise an error like `ValueError` here, catch the ValueError, and then catch other exceptions in a more general block. Or customize your exception, `Exception("No features found")` so that you have a clearer idea of where the code breaks if it breaks.

Reply:
Yeah this looks great. Thank you :slightly_smiling_face: